tirocks/src/option/cf.rs

```diff
@@ -0,0 +1,1279 @@
+        sample_file_size_ratio: f64
+
+        /// The blob file size less than this option will be mark GC.
+        ///
+        /// Default: 8MB
+        merge_small_file_threshold: u64
+
+        /// The mode used to process blob file.
+        ///
+        /// Default: kNormal
+        blob_run_mode: TitanBlobRunMode
+
+        /// If set true, values in blob file will be merged to a new blob file while
+        /// their corresponding keys are compacted to last two level in LSM-Tree.
+        ///
+        /// With this feature enabled, Titan could get better scan performance, and
+        /// better write performance during GC, but will suffer around 1.1 space
+        /// amplification and 3 more write amplification if no GC needed (eg. uniformly
+        /// distributed keys) under default rocksdb setting.
+        ///
+        /// Requirement: level_compaction_dynamic_level_base = true
+        /// Default: false
+        level_merge: bool
+
+        /// With level merge enabled, we expect there are no more than 10 sorted runs
+        /// of blob files in both of last two levels. But since last level blob files
+        /// won't be merged again, sorted runs in last level will increase infinitely.
+        ///
+        /// With this feature enabled, Titan will check sorted runs of compaction range
+        /// after each last level compaction and mark related blob files if there are
+        /// too many. These marked blob files will be merged to a new sorted run in
+        /// next compaction.
+        ///
+        /// Default: false
+        range_merge: bool
+
+        /// Max sorted runs to trigger range merge. Decrease this value will increase
+        /// write amplification but get better short range scan performance.
+        ///
+        /// Default: 20
+        max_sorted_runs: i32
+
+        /// If set true, Titan will rewrite valid blob index from GC output as merge
+        /// operands back to data store.
+        ///
+        /// With this feature enabled, Titan background GC won't block online write,
+        /// trade-off being read performance slightly reduced compared to normal
+        /// rewrite mode.
+        ///
+        /// Default: false
+        gc_merge_rewrite: bool
+    }
+
+    pub(crate) unsafe fn from_ptr<'a>(ptr: *const rocksdb_titandb_TitanCFOptions) -> &'a Self {
+        &*(ptr as *const RawTitanCfOptions)
+    }
+
+    pub(crate) unsafe fn from_ptr_mut<'a>(
+        ptr: *mut rocksdb_titandb_TitanCFOptions,
+    ) -> &'a mut Self {
+        &mut *(ptr as *mut RawTitanCfOptions)
+    }
+
+    pub(crate) fn as_mut_ptr(&mut self) -> *mut rocksdb_titandb_TitanCFOptions {
+        self as *mut _ as _
+    }
+
+    pub(crate) fn as_ptr(&self) -> *const rocksdb_titandb_TitanCFOptions {
+        self as *const _ as _
+    }
+}
+
+impl TitanCfOptions {
+    /// Same as `CfOptions::set_comparator`.
+    #[inline]
+    pub fn set_comparator(&mut self, c: Arc<SysComparator>) -> &mut Self {
```

Comment:
It doesn't look very consistent that comparator can be set directly on a Titan options, but other RocksDB options cannot. I suggest that `TitanCfOptions` ctor requires a `CfOptions`, so we don't need to worry about these RocksDB configurations here.

Reply:
> but other RocksDB options cannot

`TitanCfOptions` deref to `CfOptions`, so any options can be set to `CfOptions` can also be set to `TitanCfOptions` directly.

Reply:
No, it derefs to `RawTitanCfOptions` and `RawCfOptions` but not `CfOptions`.

Reply:
Yes. All rocksdb options are implemented for `RawCfOptions`. `TitanCfOptions` deref to `RawTitanCfOptions`, `RawTitanCfOptions` deref to `RawCfOptions`, `CfOptions` deref to `RawCfOptions`. So All methods of `RawCfOptions` can be called on `CfOptions`, `RawTitanCfOptions` and `TitanCfOptions`. Methods only belong to `CfOptions` is setting comparator, that's why it's also implemented for `TitanCfOptions`.

Reply:
Okay.